src/hooks/simulation/useGetDifference.ts

```diff
@@ -0,0 +1,80 @@
+import { useCurrentSimulation, useFormState, useRule } from '@/publicodes-state'
+import type { Metrics } from '@incubateur-ademe/nosgestesclimat'
+import { useEffect, useRef, useState } from 'react'
+
+export function useGetDifference({ metric }: { metric: Metrics }): {
+  difference: number
+  updateKey: number
+} {
+  const { numericValue } = useRule('bilan', metric)
+
+  const [difference, setDifference] = useState(0)
+  const [countValueUpdate, setCountValueUpdate] = useState(0)
+  // We need this value to force the component to re-render when the numericValue changes
+  // We don't use numericValue directly because it update before the displayDifference
+  const [keyFromNumericValue, setKeyFromNumericValue] = useState(numericValue)
+  const [hasAlreadyAnswered, setHasAlreadyAnswered] = useState<
+    boolean | undefined
+  >(undefined)
+
+  const { foldedSteps } = useCurrentSimulation()
+
+  const { currentQuestion } = useFormState()
+
+  const prevValue = useRef(numericValue)
+  const prevQuestion = useRef(currentQuestion)
+
+  // Reset component state upon question change
+  useEffect(() => {
+    if (prevQuestion.current !== currentQuestion) {
+      setDifference(0)
+      setHasAlreadyAnswered(undefined)
+      setCountValueUpdate(0)
+    }
+  }, [currentQuestion])
+
+  // Dynamically set hasAlreadyAnswered when user visits question he/she has already answered
+  useEffect(() => {
+    const isInFoldedSteps = !!(
+      foldedSteps &&
+      currentQuestion &&
+      foldedSteps.includes(currentQuestion)
+    )
+
+    if (typeof hasAlreadyAnswered === 'undefined') {
+      setHasAlreadyAnswered(isInFoldedSteps)
+    }
+  }, [currentQuestion, foldedSteps, hasAlreadyAnswered])
+
+  // If user has already answered, show the value change upon next form interaction
+  useEffect(() => {
+    if (hasAlreadyAnswered && countValueUpdate === 0) {
+      setCountValueUpdate(1)
+    }
+  }, [countValueUpdate, hasAlreadyAnswered, difference])
+
+  // Calculate difference
+  useEffect(() => {
+    const updatedDifference = numericValue - prevValue.current
+
+    setDifference(updatedDifference)
+
+    setKeyFromNumericValue(numericValue)
+
+    if (numericValue !== prevValue.current) {
+      setCountValueUpdate((prevValue) => ++prevValue)
+    }
+
+    prevValue.current = numericValue
+  }, [numericValue])
```

Comment:
Maybe you could try to merge the 3 useEffects in one using state object but that's OK

Reply:
De mon expérience il vaut mieux créer de multiples useEffect n'ayant chacun que quelques dépendances, sans quoi on s'expose à des effets de bord pas piqué des hannetons